### PR TITLE
Combine account locking and block space reserving in single loop

### LINKED ISSUE #34825: Transaction locking fails prematurely due to cost tracker constraints
#### Problem
When selecting transactions from a batch for execution in banking stage, there are two preliminary steps: qos cost reservation and global lock acquisition. In the first step, we iterate over a batch of transactions and reserve qos service cost capacity until capacity is depleted. Any transactions towards the end of the batch list that cannot fit, will be marked for retry and skipped for the current batch. Then, in the next step, we again iterate over the batch of transactions and attempt to acquire locks for any un-skipped transactions. If any locks fail, there is leftover cost capacity that could be used for the previously skipped transactions.

Credit: @crispheaney

#### Proposed Solution
Combine the loops for qos cost reservation and global lock acquisition and use the following process:
1. Reserve transaction cost in qos service
2. Attempt to acquire locks
3. If lock acquisition fails, unreserve transaction cost in qos service
4. Continue to next transaction

cc @taozhu-chicago @apfitzge

### PR DESCRIPTION
**Pull request by tao-stones**
#### Problem

Transaction takes up block space before acquiring accounts lock, if it fails getting lock, block space is not immediately released.

#### Summary of Changes
- release reserved block space immediately after lock acquisition 
- replace transaction's cached cost with account lock error if block space released, this is to prevent possibility of double-removing its costs.

Fixes #34825

### CODE REVIEW

File: core/src/banking_stage/consumer.rs [hunk 1]
```diff
@@ -479,6 +479,32 @@ impl Consumer {
...
+                    if cost.is_ok() {
+                        // set cost to lock_err, so it won't be  accidentally removed more than once
+                        let mut c = Err(lock_err.clone());
+                        std::mem::swap(cost, &mut c);
+                        return Some(c.unwrap());
+                    }
+                }
+                None
+            })
+            .collect();
```
Comment (apfitzge) [inline core/src/banking_stage/consumer.rs:498]: we don't need to allocate a new vec here, we can just iterate over our qos and lock result pairs...something like
```rust
        // Remove reserved block space for transaction failed to acquire lock
        let mut cost_tracker = bank.write_cost_tracker().unwrap();
        transaction_qos_cost_results
            .iter_mut()
            .zip(batch.lock_results())
            .filter_map(|(qos_cost_result, lock_result)| {
                match (qos_cost_result.as_ref(), lock_result) {
                    (Ok(_), Err(lock_err)) => Some((qos_cost_result, lock_err)),
                    _ => None,
                }
            })
            .for_each(|(qos_cost_result, lock_err)| {
                cost_tracker.remove(qos_cost_result.as_ref().unwrap());
                *qos_cost_result = Err(lock_err.clone());
            });
```

File: core/src/banking_stage/consumer.rs [hunk 2]
```diff
@@ -460,7 +460,7 @@ impl Consumer {
         pre_results: impl Iterator<Item = Result<(), TransactionError>>,
     ) -> ProcessTransactionBatchOutput {
         let (
-            (transaction_qos_cost_results, cost_model_throttled_transactions_count),
+            (mut transaction_qos_cost_results, cost_model_throttled_transactions_count),
```
Comment (apfitzge) [inline core/src/banking_stage/consumer.rs:463]: as an alternative to mutability here. we could also use the lock-results to inform the cost-adjustment post-execution. Not sure if that would be cleaner though.

File: core/src/banking_stage/consumer.rs (hunk 1 above)
Comment (tao-stones) [reply core/src/banking_stage/consumer.rs:498]: TIL! I was having trouble to change `qos_cost_result` while it is borrowed by `match` (i think).

Before making the change tho, are we good with this approach, or would like to "lock-reserve-unlock" you mentioned elsewhere?

**Comment by willhickey** [comment]
This repository is no longer in use. Please re-open this pull request in the agave repo: https://github.com/anza-xyz/agave

**Pull request closed without merging**